File: src/main/java/io/codeka/gaia/config/security/oauth2/OAuth2SuccessHandler.java
package io.codeka.gaia.config.security.oauth2;

import io.codeka.gaia.config.security.SuccessHandler;
import io.codeka.gaia.registries.RegistryOAuth2Provider;
import io.codeka.gaia.teams.OAuth2User;
import io.codeka.gaia.teams.User;
import io.codeka.gaia.teams.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.security.oauth2.client.ClientsConfiguredCondition;
import org.springframework.context.annotation.Conditional;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
import org.springframework.security.oauth2.core.user.DefaultOAuth2User;
import org.springframework.stereotype.Component;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.List;

@Component
@Conditional(ClientsConfiguredCondition.class)
public class OAuth2SuccessHandler extends SuccessHandler {

    private List<RegistryOAuth2Provider> registryOAuth2Providers;
    private OAuth2AuthorizedClientService oAuth2AuthorizedClientService;

    @Autowired
    public OAuth2SuccessHandler(UserRepository userRepository, List<RegistryOAuth2Provider> registryOAuth2Providers,
                                OAuth2AuthorizedClientService oAuth2AuthorizedClientService) {
        super(userRepository);
        this.registryOAuth2Providers = registryOAuth2Providers;
        this.oAuth2AuthorizedClientService = oAuth2AuthorizedClientService;
    }

    @Override
    public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response, Authentication authentication) throws IOException {
        // get user if exist, otherwise create a new one
        var user = userRepository.findById(authentication.getName())
                .orElse(new User(authentication.getName(), null));
        // get oauth2 data
        user.setOAuth2User(getOAuth2User((OAuth2AuthenticationToken) authentication));
        userRepository.save(user);
        redirect(request, response);
    }

    private OAuth2User getOAuth2User(OAuth2AuthenticationToken authentication) {
        var client = oAuth2AuthorizedClientService.loadAuthorizedClient(
                authentication.getAuthorizedClientRegistrationId(),
                authentication.getName());
        var user = (DefaultOAuth2User) authentication.getPrincipal();
        return registryOAuth2Providers.stream()
                .filter(p -> p.isAssignableFor(client.getClientRegistration().getRegistrationId()))
                .map(p -> p.getOAuth2User(user, client))
                .findFirst().orElse(null);
    }
}

File: src/main/java/io/gaia_app/config/security/oauth2/OAuth2SuccessHandler.kt
package io.gaia_app.config.security.oauth2

import io.gaia_app.config.security.SuccessHandler
import io.gaia_app.registries.RegistryOAuth2Provider
import io.gaia_app.organizations.OAuth2User
import io.gaia_app.organizations.User
import io.gaia_app.organizations.repository.UserRepository
import org.springframework.boot.autoconfigure.security.oauth2.client.ClientsConfiguredCondition
import org.springframework.context.annotation.Conditional
import org.springframework.security.core.Authentication
import org.springframework.security.oauth2.client.OAuth2AuthorizedClient
import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken
import org.springframework.security.oauth2.core.user.DefaultOAuth2User
import org.springframework.stereotype.Component
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
@Conditional(ClientsConfiguredCondition::class)
class OAuth2SuccessHandler(
        userRepository: UserRepository,
        val registryOAuth2Providers: List<RegistryOAuth2Provider>,
        val oAuth2AuthorizedClientService: OAuth2AuthorizedClientService) : SuccessHandler(userRepository) {

    override fun onAuthenticationSuccess(request: HttpServletRequest, response: HttpServletResponse, authentication: Authentication) {
        // get user if exist, otherwise create a new one
        val user = userRepository
            .findById(authentication.name)
            .orElse(User(authentication.name, null))

        // get oauth2 data
        user.oAuth2User = getOAuth2User(authentication as OAuth2AuthenticationToken)
        userRepository.save(user)

        response.sendRedirect("/")
    }

    private fun getOAuth2User(authentication: OAuth2AuthenticationToken): OAuth2User? {
        val client = oAuth2AuthorizedClientService.loadAuthorizedClient<OAuth2AuthorizedClient>(
            authentication.authorizedClientRegistrationId,
            authentication.name)

        return registryOAuth2Providers
            .filter { it.isAssignableFor(client.clientRegistration.registrationId) }
            .map { it.getOAuth2User(authentication.principal as DefaultOAuth2User, client) }
            .firstOrNull()
    }

}
